modules/seqexec/web/client/src/main/scala/seqexec/web/client/components/sequence/steps/NodAndShuffleProgress.scala

```diff
@@ -0,0 +1,266 @@
+            ongoingSectionColor = if (isInError) "red".some else "blue".some,
+            progressCls = List(SeqexecStyles.observationProgressBar),
+            barCls = List(SeqexecStyles.observationBar),
+            labelCls = List(SeqexecStyles.observationLabel),
+            stopping = p.isStopping,
+            paused = p.summary.step.isObservePaused
+            )
+        } getOrElse
+          <.div("NodAndShuffleProgress invoked without a Nod&Shuffle step summary")
+      }
+    }
+    .configure(Reusability.shouldComponentUpdate)
+    .build
+}
+
+final case class NodAndShuffleCycleProgress(summary: StepStateSummary) extends ReactProps {
+  @inline def render: VdomElement = NodAndShuffleCycleProgress.component(this)
+}
+
+object NodAndShuffleCycleProgress {
+  type Props = NodAndShuffleCycleProgress
+
+  implicit val propsReuse: Reusability[Props] = Reusability.derive[Props]
+
+  protected val component = ScalaComponent
+    .builder[Props]("NodAndShuffleCycleProgress")
+    .stateless
+    .render_P { p =>
+      NodAndShuffleProgress(
+        p.summary,
+        nsStatus => List.range(1, nsStatus.cycles + 1).map(_.show),
+        (nsStatus, nodMillis) => {
+          val cycleMillis = nodMillis * NodAndShuffleStage.NsSequence.length
+          val currentValue = nsStatus.state.map { s =>
+            s.sub.cycle * cycleMillis + s.sub.stageIndex * nodMillis
+          }
+          (cycleMillis, currentValue)
+        }
+      )
+    }
+    .configure(Reusability.shouldComponentUpdate)
+    .build
+}
+
+final case class NodAndShuffleNodProgress(summary: StepStateSummary) extends ReactProps {
+  @inline def render: VdomElement = NodAndShuffleNodProgress.component(this)
+}
+
+object NodAndShuffleNodProgress {
+  type Props = NodAndShuffleNodProgress
+
+  implicit val propsReuse: Reusability[Props] = Reusability.derive[Props]
+
+  protected val component = ScalaComponent
+    .builder[Props]("NodAndShuffleNodProgress")
+    .stateless
+    .render_P { p =>
+      NodAndShuffleProgress(
+        p.summary,
+        _ => NodAndShuffleStage.NsSequence.map(_.symbol.name).toList,
+        (nsStatus, nodMillis) => {
+          val currentValue = nsStatus.state.map(_.sub.stageIndex * nodMillis)
+          (nodMillis, currentValue)
+        }
+        )
+    }
+    .configure(Reusability.shouldComponentUpdate)
+    .build
+}
+
+sealed trait NodAndShuffleRowProps extends ReactProps {
+  val clientStatus: ClientStatus
+  val stateSummary: StepStateSummary
+}
+
+trait NodAndShuffleRow[L <: OperationLevel] {
+  type Props <: NodAndShuffleRowProps
+
+  implicit val propsReuse: Reusability[Props]
+
+  implicit val propsControlButtonResolver: ControlButtonResolver[Props] =
+    ControlButtonResolver.build(p => (p.clientStatus, p.stateSummary.state, p.stateSummary.step))
+
+  implicit private val operationLevelType: OperationLevelType[L] = implicitly[OperationLevelType[L]]

```

Comment:
Who uses this?

Reply:
It's passed into `SupportedOperationsOps.operations` in order to be able to `match` on the type parameter `L <: OperationLevel`. I accept suggestions as to other ways of doing this.

BTW, I had to move the implicit resolutions into the subclasses of `NodAndShuffleRow`, that's why the buttons were not being displayed.

Reply:
Hmm.... I think I'm going to try to create a 2-parameter typeclass depending on the `[Instrument, OperationLevel]` and simplify the whole logic.